x/tradeshield/keeper/pending_perpetual_order.go

```diff
@@ -67,25 +67,45 @@ func (k Keeper) AppendPendingPerpetualOrder(
 func (k Keeper) SetPendingPerpetualOrder(ctx sdk.Context, pendingPerpetualOrder types.PerpetualOrder) {
 	store := prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(ctx)), types.PendingPerpetualOrderKey)
 	b := k.cdc.MustMarshal(&pendingPerpetualOrder)
-	store.Set(GetPendingPerpetualOrderIDBytes(pendingPerpetualOrder.OrderId), b)
+	store.Set(types.GetPendingPerpetualOrderKeyBytes(sdk.MustAccAddressFromBech32(pendingPerpetualOrder.OwnerAddress), pendingPerpetualOrder.PoolId, pendingPerpetualOrder.OrderId), b)
 }
 
 // GetPendingPerpetualOrder returns a pendingPerpetualOrder from its id
-func (k Keeper) GetPendingPerpetualOrder(ctx sdk.Context, id uint64) (val types.PerpetualOrder, found bool) {
+func (k Keeper) GetPendingPerpetualOrder(ctx sdk.Context, user sdk.AccAddress, poolId uint64, orderId uint64) (val types.PerpetualOrder, found bool) {
 	store := prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(ctx)), types.PendingPerpetualOrderKey)
-	b := store.Get(GetPendingPerpetualOrderIDBytes(id))
+	b := store.Get(types.GetPendingPerpetualOrderKeyBytes(user, poolId, orderId))
 	if b == nil {
 		return val, false
 	}
 	k.cdc.MustUnmarshal(b, &val)
 	return val, true
 }
 
+// DeletePendingPerpetualOrdersForAddressAndPool deletes all pending perpetual orders for a given address, pool id and position id
+func (k Keeper) DeletePendingPerpetualOrdersForAddressAndPool(ctx sdk.Context, user sdk.AccAddress, poolId uint64, positionId uint64) ([]types.PerpetualOrder, error) {
+	store := prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(ctx)), types.PendingPerpetualOrderKey)
+	iterator := storetypes.KVStorePrefixIterator(store, types.GetPendingPerpetualOrderAddressPoolKey(user, poolId))
+	defer iterator.Close()
+
+	var orders []types.PerpetualOrder
+	for ; iterator.Valid(); iterator.Next() {
+		var order types.PerpetualOrder
+		k.cdc.MustUnmarshal(iterator.Value(), &order)
+		if order.Status == types.Status_PENDING && order.PositionId == positionId &&
+			(order.PerpetualOrderType == types.PerpetualOrderType_LIMITCLOSE || order.PerpetualOrderType == types.PerpetualOrderType_STOPLOSSPERP) {
+			store.Delete(types.GetPendingPerpetualOrderKeyBytes(user, poolId, order.OrderId))
```

Comment:
can we emit event for each order getting deleted / cancelled, will be useeful for ccxt

Reply:
yeah, will check open too, events are missing